src/Shell/Task/ModelTask.php

```diff
@@ -344,25 +344,56 @@ public function findHasMany($model, array $associations)
                 }
             }
 
-            foreach ($otherSchema->columns() as $fieldName) {
-                $assoc = false;
-                if (!in_array($fieldName, $primaryKey) && $fieldName === $foreignKey) {
-                    $assoc = [
-                        'alias' => $otherModel->alias(),
-                        'foreignKey' => $fieldName
-                    ];
-                } elseif ($otherTableName === $tableName && $fieldName === 'parent_id') {
-                    $className = ($this->plugin) ? $this->plugin . '.' . $model->alias() : $model->alias();
-                    $assoc = [
-                        'alias' => 'Child' . $model->alias(),
-                        'className' => $className,
-                        'foreignKey' => $fieldName
-                    ];
+            $targetForeignKey = $primaryKey;
+            foreach ($targetForeignKey as $i => $fieldName) {
+                if ($fieldName === 'id') {
+                    $targetForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
                 }
-                if ($assoc && $this->plugin && empty($assoc['className'])) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+            }
+
+            $possibleForeignKeys = $otherSchema->columns();
+            foreach ($possibleForeignKeys as $i => $fieldName) {
+                if ($fieldName === 'parent_id') {
+                    continue;
+                }
+                if (strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName)) {
+                    unset($possibleForeignKeys[$i]);
                 }
-                if ($assoc) {
+            }
+
+            $assoc = false;
+            if (Hash::diff(
+                array_values($targetForeignKey), array_values($possibleForeignKeys)) === []
```

Comment:
I wanted to say that I don't think this `Hash::diff()` call works as you expected. In the above example, wouldn't `$possibleForeginKeys` be `['site_id', 'article_id', 'user_id']`? Then I think that calling `Hash::diff()` would never return `[]`.